crates/goose/src/agents/todo_extension.rs

```diff
@@ -0,0 +1,263 @@
+                        .extension_data(session.extension_data)
+                        .apply()
+                        .await
+                    {
+                        Ok(_) => Ok(vec![Content::text(format!(
+                            "Updated ({} chars)",
+                            char_count
+                        ))]),
+                        Err(_) => Err("Failed to update session metadata".to_string()),
+                    }
+                } else {
+                    Err("Failed to serialize TODO state".to_string())
+                }
+            }
+            Err(_) => Err("Failed to read session metadata".to_string()),
+        }
+    }
+
+    fn get_tools() -> Vec<Tool> {
+        fn create_schema(json_value: Value) -> Arc<Map<String, Value>> {
+            Arc::new(json_value.as_object().unwrap().clone())
+        }
+
+        vec![
+            Tool {
+                name: "todo_read".into(),
+                description: Some(indoc! {r#"
+                Read the entire TODO file content.
+
+                This tool reads the complete TODO file and returns its content as a string.
+                Use this to view current tasks, notes, and any other information stored in the TODO file.
+
+                The tool will return an error if the TODO file doesn't exist or cannot be read.
+            "#}.into()),
+                input_schema: create_schema(json!({
+                "type": "object",
+                "properties": {},
+                "required": []
+            })),
+                annotations: None,
+                output_schema: None,
+            },
+            Tool {
+                name: "todo_write".into(),
+                description: Some(indoc! {r#"
+                Write or overwrite the entire TODO file content.
+
+                This tool replaces the complete TODO file content with the provided string.
+                Use this to update tasks, add new items, or reorganize the TODO file.
+
+                WARNING: This operation completely replaces the file content. Make sure to include
+                all content you want to keep, not just the changes.
+
+                The tool will create the TODO file if it doesn't exist, or overwrite it if it does.
+                Returns an error if the file cannot be written due to permissions or other I/O issues.
+            "#}.into()),
+                input_schema: create_schema(json!({
+                "type": "object",
+                "properties": {
+                    "content": {
+                        "type": "string",
+                        "description": "The TODO list content to save"
+                    }
+                },
+                "required": ["content"]
+            })),
+                annotations: None,
+                output_schema: None,
+            },
+        ]
+    }
+}
+
+#[async_trait]
+impl McpClientTrait for TodoClient {
+    async fn list_resources(
```

Comment:
if these are going to act like mcp servers, it would be really nice to be able to define them using rmcp's server sdk, e.g.

```
impl Todo {
    #[tool(name = "todo_add")]
    pub async fn todo_add(...)
}
```

they could still be built with the platform extension context and have all the same access, but they'd be easier to write and easily port-able to out-of-process MCP servers